src/fam.rs

```diff
@@ -0,0 +1,623 @@
+        }
+
+        Ok(())
+    }
+
+    /// Appends an element to the end of the collection and updates `len`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The element that will be appended to the end of the collection.
+    ///
+    /// # Error: When len is already equal to max possible len it returns Error::SizeLimitExceeded
+    ///
+    pub fn push(&mut self, entry: T::Entry) -> Result<(), Error> {
+        let new_len = self.len() + 1;
+        self.set_len(new_len)?;
+        self.as_mut_slice()[new_len - 1] = entry;
+
+        Ok(())
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// # Arguments
+    ///
+    /// * `f` - The function used to evaluate whether an entry will be kept or not.
+    ///         When `f` returns `true` the entry is kept.
+    ///
+    pub fn retain<P>(&mut self, mut f: P)
+    where
+        P: FnMut(&T::Entry) -> bool,
+    {
+        let mut num_kept_entries = 0;
+        {
+            let entries = self.as_mut_slice();
+            for entry_idx in 0..entries.len() {
+                let keep = f(&entries[entry_idx]);
+                if keep {
+                    entries[num_kept_entries] = entries[entry_idx];
+                    num_kept_entries += 1;
+                }
+            }
+        }
+
+        // This is safe since this method is not increasing the len
+        self.set_len(num_kept_entries).expect("invalid length");
+    }
+}
+
+impl<T: Default + FamStruct> PartialEq for FamStructWrapper<T> {
+    fn eq(&self, other: &FamStructWrapper<T>) -> bool {
+        self.as_slice() == other.as_slice()
+    }
+}
+
+impl<T: Default + FamStruct + Clone> Clone for FamStructWrapper<T> {
+    fn clone(&self) -> Self {
+        FamStructWrapper {
+            mem_allocator: self.mem_allocator.to_vec(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const MAX_LEN: usize = 100;
+
+    #[repr(C)]
+    #[derive(Default)]
+    pub struct __IncompleteArrayField<T>(::std::marker::PhantomData<T>, [T; 0]);
+    impl<T> __IncompleteArrayField<T> {
+        #[inline]
+        pub fn new() -> Self {
+            __IncompleteArrayField(::std::marker::PhantomData, [])
+        }
+        #[inline]
+        pub unsafe fn as_ptr(&self) -> *const T {
+            ::std::mem::transmute(self)
+        }
+        #[inline]
+        pub unsafe fn as_mut_ptr(&mut self) -> *mut T {
+            ::std::mem::transmute(self)
+        }
+        #[inline]
+        pub unsafe fn as_slice(&self, len: usize) -> &[T] {
+            ::std::slice::from_raw_parts(self.as_ptr(), len)
+        }
+        #[inline]
+        pub unsafe fn as_mut_slice(&mut self, len: usize) -> &mut [T] {
+            ::std::slice::from_raw_parts_mut(self.as_mut_ptr(), len)
+        }
+    }
+
+    impl<T> ::std::clone::Clone for __IncompleteArrayField<T> {
+        #[inline]
+        fn clone(&self) -> Self {
+            Self::new()
+        }
+    }
+
+    #[repr(C)]
```

Comment:
I would put the following lines (till 398) as a documentation example for the fam module.